storage/local/frankenstein.go

```diff
@@ -259,3 +331,133 @@ func (i *Ingestor) Collect(ch chan<- prometheus.Metric) {
 	ch <- i.storedChunks
 	ch <- i.chunkStoreFailures
 }
+
+type invertedIndex struct {
+	mtx sync.RWMutex
+	idx map[model.LabelName]map[model.LabelValue][]model.Fingerprint // entries are sorted in fp order?
+}
+
+func newInvertedIndex() *invertedIndex {
+	return &invertedIndex{
+		idx: map[model.LabelName]map[model.LabelValue][]model.Fingerprint{},
+	}
+}
+
+func (i *invertedIndex) add(metric model.Metric, fp model.Fingerprint) {
+	i.mtx.Lock()
+	defer i.mtx.Unlock()
+
+	for name, value := range metric {
+		values, ok := i.idx[name]
+		if !ok {
+			values = map[model.LabelValue][]model.Fingerprint{}
+		}
+		fingerprints := values[value]
+		j := sort.Search(len(fingerprints), func(i int) bool {
+			return fingerprints[i] >= fp
+		})
+		fingerprints = append(fingerprints, 0)
+		copy(fingerprints[j+1:], fingerprints[j:])
+		fingerprints[j] = fp
+		values[value] = fingerprints
+		i.idx[name] = values
+	}
+}
+
+func (i *invertedIndex) lookup(matchers []*metric.LabelMatcher) []model.Fingerprint {
+	if len(matchers) == 0 {
+		return nil
+	}
+	i.mtx.RLock()
+	defer i.mtx.RUnlock()
+
+	// intersection is initially nil, which is a special case.
+	var intersection []model.Fingerprint
+	for _, matcher := range matchers {
+		values, ok := i.idx[matcher.Name]
+		if !ok {
+			return nil
+		}
+		var toIntersect []model.Fingerprint
+		for value, fps := range values {
+			if matcher.Match(value) {
+				toIntersect = union(toIntersect, fps)
```

Comment:
Is an explicit `union` here necessary rather than just an append? There shouldn't be any dupes within this inner loop because with different label values for the same label name, the fingerprint sets will always be disjoint.